include/GMRESSolver.hpp

```diff
@@ -0,0 +1,352 @@
+    return std::any_of(x.begin(), x.end(),
+        [](real_type val) { return val != real_type(0); });
+}
+} // namespace details
+
+// Convergence flags
+// unconverged: neither convergence nor divergence detected
+// success:  residual norm <= convTol
+// diverged: residual norm >= divTol
+// maxits: maximum number of iterations were reached 
+// true_rnrm: true residual norm > convTol
+// inf_or_nan: residual norm is inf or nan
+// precon_fail: application of the preconditioner failed
+// matvec_fail: matrix-vector product failed
+// breakdown: input matrix or preconditioner are (likely) singular
+// stagnation: |xnew(i) - x(i)| <= stgTol * |xnew(i)| for all i
+enum class ConvFlag { unconverged, success, diverged, maxits, 
+    true_rnrm, inf_or_nan, precon_fail, matvec_fail, breakdown,
+    stagnation };
+
+// convert a ConvFlag to a string
+std::string ConvFlagToStr(ConvFlag flag);
+
+inline std::ostream& operator<<(std::ostream& strm, ConvFlag flag) {
+    strm << ConvFlagToStr(flag);
+    return strm;
+}
+
+struct GMRESParams {
+    bool verbose;        // if true, output convergence history to console 
+    bool testTrueRnrm;   // if true, check that true residual norm of solution <= convTol
+    index_type kspaceSz; // max size of Krylov subspace before restarting
+    index_type maxits;   // max number of outer iterations
+    real_type relTol;    // relative convergence tolerance
+    real_type absTol;    // absolute convergence tolerance
+    real_type divTol;    // divergence tolerance
+    real_type stgTol;    // stagnation tolerance
+    GMRESParams()
+        : verbose{ false }, testTrueRnrm{ true }, 
+        kspaceSz{ 30 }, maxits{ 100 }, relTol{ 1e-6 }, 
+        absTol{ 1e-6 }, divTol{ 1e5 }, stgTol{ 1e-12 }
+    {}
+};
+
+// throws an exeception if any parameters in p are out of range
+void checkGMRESParams(const GMRESParams& p);
+
+struct GMRESOut {
+    ConvFlag flag;       // convergence flag
+    index_type innerIts; // inner iteration number at which x was computed
+    index_type outerIts; // outer iteration number at which x was computed
+    real_type relres;    // scaled residual norm of solution
+    std::string msg;     // addtional output info
+    GMRESOut()
+        : flag{ ConvFlag::unconverged },
+        innerIts{ 0 }, outerIts{ 0 }, relres{ real_type(0) },
+        msg{}
+    {}
+};
+
+std::ostream& operator<<(std::ostream& strm, const GMRESOut& out);
+
+// Implements preconditioned, restarted GMRES with Modified
+// Gram-Schmidt orthogonalization. Solves the preconditioned
+// linear system A*M^{-1}*u = b, where u = M*x and
+// M is the preconditioner. The solver converges at the kth
+// iteration if
+//     1. k < maxits * kspaceSz
+//     2. ||r_k|| < divTol * ||r_0||
+//     3. ||r_k|| <= max(relTol * ||b||, absTol)
+// where ||.|| is the vector 2-norm and r_k is the 
+// unpreconditioned residual of the kth iterate.
+class GMRESSolver {
+public:
+    // Input/Output:
```

Comment:
Nice documentation, can we change the commenting syntax to use the /**  */ style used elsewhere in the project? Otherwise doxygen won't know to pull these into the docs

Reply:
Roger that Frank, I will make these changes.